elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/annotation/SQLExpression.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2019, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.datastores.aggregation.queryengines.sql.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Provide arbitrary sql expression on dimension field or method.
+ * This expression should use '%reference' as the place holder for physical table column reference.
+ * i.e. "IF(%reference >= 0, 'positive', 'negative')".
+ *
+ * Sql expression can only be applied on degenerated dimension field or physical column reference.
+ * Using this annotation on metric field is invalid as metric function expression already has the same functionality.
+ * Using this annotation on relationship field is invalid as relationship field is not physical table column.
+ * Sql expression can be carried when used with @JoinTo, the outer layer expression would be applied after the initial
+ * joined to sql expression.
+ */
+@Documented
+@Target({ElementType.FIELD, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface SQLExpression {
```

Comment:
I think we should:
1. Unify this concept with MetricComputation.  They are closely related.
2. Instead of arbitrary names like 'reference', we should use model field names like %overallRating and %highScore.   This will allow metric definitions as well.

We still need to differentiate whether something is a metric or dimension though.

Reply:
My initial thought is that MetricComputation is resolved as MetricFunction, SQLExpression is for dimension fields. We would not use dimension fields for metric computation, right?

Reply:
We can support `@SQLExpression` on metric field, like
```java
@SQLExpression("%reference / 1000")
@MetricAggregation(function = SqlMax.class)
@Column(name = "col1")
private float highScore;
```
Would be resolved as `SELECT MAX(col / 1000) AS highScore`.
I idea is to keep `SQLExpression` separated from metrics, as metrics would be loaded into metadata store to provide information like arguments, functions and so on.